core/blockchain.go

```diff
@@ -275,18 +234,19 @@ func (bc *BlockChain) SetHead(head uint64) {
 	if bc.currentBlock == nil {
 		bc.currentBlock = bc.genesisBlock
 	}
-	if bc.currentHeader == nil {
-		bc.currentHeader = bc.genesisBlock.Header()
-	}
 	if bc.currentFastBlock == nil {
 		bc.currentFastBlock = bc.genesisBlock
 	}
+	if bc.hc.CurrentHeader().Number.Uint64() < bc.currentBlock.NumberU64() {
+		bc.currentBlock = bc.GetBlock(bc.hc.CurrentHeader().Hash())
+	}
+	if bc.hc.CurrentHeader().Number.Uint64() < bc.currentFastBlock.NumberU64() {
+		bc.currentFastBlock = bc.GetBlock(bc.hc.CurrentHeader().Hash())
+	}
```

Comment:
Please move these two ifs to before the nil checks. As you are only modifying the `currentBlock` and `currentFastBlock` here, they cannot be nil before, however they **can** become nil with these assignments since the header may be ahead of the blocks, in case both will be nil after a `SetHead`.